editors/code/src/cargo.ts

```diff
@@ -0,0 +1,109 @@
+                message => {
+                    if (message.reason == 'compiler-artifact' && message.executable) {
+                        let isBinary = message.target.crate_types.includes('bin');
+                        let isBuildScript = message.target.kind.includes('custom-build');
+                        if ((isBinary && !isBuildScript) || message.profile.test) {
+                            artifacts.push({
+                                fileName: message.executable,
+                                name: message.target.name,
+                                kind: message.target.kind[0],
+                                isTest: message.profile.test
+                            })
+                        }
+                    }
+                    else if( message.reason == 'compiler-message') {
+                        this.output.append(message.message.rendered);
+                    }
+                },
+                stderr => {
+                    this.output.append(stderr);
+                }
+            );
+        }
+        catch (err) {
+            this.output.show(true);
+            throw new Error(`Cargo invocation has failed: ${err}`);
+        }
+
+        return artifacts;
+    }
+
+    public async executableFromArgs(cargoArgs: string[], extraArgs?: string[]): Promise<string> {
+        cargoArgs.push("--message-format=json");
+        if (extraArgs) {
+            cargoArgs.push('--');
+            cargoArgs.push(...extraArgs);
+        }
+
+        let artifacts = await this.artifactsFromArgs(cargoArgs);
+
+        if (artifacts.length == 0 ) {
+            throw new Error('No compilation artifacts');
+        } else if (artifacts.length > 1) {
+            throw new Error('Multiple compilation artifacts are not supported.');
+        }
+
+        return artifacts[0].fileName;
+    }
+
+    runCargo(
+        cargoArgs: string[],
+        onStdoutJson: (obj: any) => void,
+        onStderrString: (data: string) => void
+    ): Promise<number> {
+        return new Promise<number>((resolve, reject) => {
+            let cargo = cp.spawn('cargo', cargoArgs, {
+                stdio: ['ignore', 'pipe', 'pipe'],
+                cwd: this.rootFolder,
+                env: this.env,
+            });
+
+            cargo.on('error', err => {
+                reject(new Error(`could not launch cargo: ${err}`));
+            });
+            cargo.stderr.on('data', chunk => {
+                onStderrString(chunk.toString());
+            });
+
+            let rl = readline.createInterface({ input: cargo.stdout });
+            rl.on('line', line => {
+                let message = JSON.parse(line);
+                onStdoutJson(message);
+            });
+
+            cargo.on('exit', (exitCode, _) => {
+                if (exitCode == 0)
+                    resolve(exitCode);
+                else
+                    reject(new Error(`exit code: ${exitCode}.`));
+            });
+        });
+    }
+}
```

Comment:
Presumably, code-lldb extension contains similar logic somewhere? I wonder if we should move it server-side, long term. Or if we should somehow add a command to cargo to do this?

Reply:
Yes, CodeLLDB has very similar class [Cargo](https://github.com/vadimcn/vscode-lldb/blob/874ac867ac413865af4fe07817ea537d253ad42f/extension/cargo.ts#L24) used for generating launch configurations for Rust and process cargo tasks.

I think this functionality is better suited for rust-analyzer. I.e. it's rust-analyzer should generate debug configurations, while CodeLLDB or MS C++ tools should be used as a debug engine only.

> I wonder if we should move it server-side, long term. Or if we should somehow add a command to cargo to do this?

Could you please clarify this point? Unless I am mistaken, rust-analyzer can find all tests\main functions, but cannot predict the executable name. So, anyway we have to call cargo to get real file names.

Reply:
I mean, we can call Cargo inside rust-analyzer server, instead of calling it in the editor plugin, like we do now. Although, given that the code would exist at least three times:

* in rust-analyzer
* in vscode-lldb
* in IntelliJ Rust

I wonder if we perhaps should upstream this to cargo somehow... There are also various finky detail like that you want to turn `cargo test`  into `cargo test --no-run`, but `cargo run` into `cargo build`...

But these all is def outside of scope for the PR, I think starting with TS impl is the right choice here.

Reply:
I get it now, thanks.

Artifacts information might be useful for IDEs that use LSP.
I do not think that cargo should do something special in this way. With `--message-format=json` it already returns all necessary information. We just need to parse it somewhere. Maybe we can add custom commands just like existing `rust-analyzer/expandMacro`?

To name a few:
`rust-analyzer/artifacts` 
`rust-analyzer/modules` 
`rust-analyzer/dependencies` 

In parallel, I'm working on Visual Studio (not VS Code) extension for Rust. And have to get all this information invoking cargo. It would be great if one can do this via an LSP server.

If you do not mind, I can start working in this direction.

Reply:
>If you do not mind, I can start working in this direction.

Yeah, this probably needs some design. Rust-analyzer is pretty build-system agnostic, and I think in the ideal world:

* if you want to know semantics of the code -> you talk to rust-analyzer
* if you want to build code -> you talk to cargo

I don't think that we should proxy cargo functionality. 

OTOH, it would be cool to have

![image](https://user-images.githubusercontent.com/1711539/80839755-25c66580-8bfc-11ea-8ff1-4041436bb999.png)

working in any editor, and the best way to achieve that seems to be via implementing an LSP extension. 

I guess, I'll be fine adding something like `rust-analyzer/modules` to rust-analyzer, with the following caveats:

* the impl should be relatively clean and mergeable without too much discussion (we have an epidemic of PRs here, so careful review bandwidth is very precious :) )
* the impl should degradate nicely if there isn't Cargo (we have altenraitve `rust-project.json` format). ideally, you should get some kind of `projctStructure` response for both flavors of projects, with additional Cargo metadata if  proejct is cargo-based
* preferably, there should be some motion on upstreaming "project structure" protocol extension to the LSP proper.

Reply:
I've been thinking about something like this (rough RA Side Panel extension sketch with [druid](https://github.com/xi-editor/druid) crate as an example):
![ra-side-panel](https://user-images.githubusercontent.com/62505555/80859700-69a48380-8c6b-11ea-82d7-c72119774481.png)

If you agree that the rust-analyzer may(should?) have such a feature I’ll open an issue for design details discussion.

Reply:
Yes, I do think we need something like this. I am not sure if we need `Modules` (they are sufficiently 1-to-1 with the file system structure), but dependencies and crates structure would be very useful. 

I have certain doubts about maintenance, as this seems to be only a nice-to-have feature, but it requires some custom client-side code. But it probably won't be too bad.